## ASSIGMENT SAMPLING- SIMRAT (102203201)

**UPLOAD DATA**

In [1]:
import pandas as pd
data=pd.read_csv("Creditcard_data.csv")
data

Time        V1        V2        V3        V4        V5        V6  \
0       0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1       0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2       1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3       1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4       2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
..    ...       ...       ...       ...       ...       ...       ...   
767   575 -0.572263  0.731748  1.541254  0.150506  1.108974  0.372152   
768   579 -1.296845 -0.511605  2.404726 -0.310762 -0.319551 -0.542842   
769   579  1.214170  0.210481  0.484651  0.479768 -0.261955 -0.527039   
770   580  1.267030 -0.071114  0.037680  0.512683  0.242392  0.705212   
771   581  1.153758  0.132273  0.382969  1.405063 -0.224287 -0.197295   

           V7        V8        V9  ...       V21       V22       V23  \
0    0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1   -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2    0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3    0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4    0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
..        ...       ...       ...  ...       ...       ...       ...   
767  1.084879 -0.146329 -0.274447  ... -0.143508 -0.107582 -0.418263   
768 -0.173310  0.260423 -1.202688  ... -0.071270 -0.161175  0.088496   
769  0.021782 -0.106888 -0.037631  ... -0.224292 -0.594609  0.159877   
770 -0.226582  0.109483  0.657565  ... -0.164468 -0.177225 -0.222918   
771  0.020653  0.029260  0.412254  ... -0.107809 -0.125231 -0.057041   

          V24       V25       V26       V27       V28  Amount  Class  
0    0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1   -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
2   -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3   -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4    0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
..        ...       ...       ...       ...       ...     ...    ...  
767 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   26.72      0  
768  0.285390  0.281069 -0.370130  0.043410  0.092318   80.00      0  
769  0.091873  0.140964  0.227406 -0.017389  0.016030    5.98      0  
770 -1.245505  0.678360  0.525059  0.002920 -0.003333   12.36      0  
771  0.073082  0.633977 -0.310685  0.033590  0.015250   13.79      0  

[772 rows x 31 columns]

In [2]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
class_count_0, class_count_1 = data['Class'].value_counts()
class_0 = data[data['Class'] == 0]
class_1 = data[data['Class'] == 1]# print the shape of the class
print('class 0:', class_0.shape)
print('class 1:', class_1.shape)



class 0: (763, 31)
class 1: (9, 31)


In [4]:
pip install imbalanced-learn

Note: you may need to restart the kernel to use updated packages.


In [5]:
from sklearn.preprocessing import StandardScaler

**NORMALIZE AMOUNT COLUMN**

In [6]:
sc= StandardScaler()
data['Amount'] = sc.fit_transform(pd.DataFrame(data['Amount']))

In [7]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28    Amount  Class  
0 -0.189115  0.133558 -0.021053  0.409447      0  
1  0.125895 -0.008983  0.014724 -0.333712      1  
2 -0.139097 -0.055353 -0.059752  1.567910      0  
3 -0.221929  0.062723  0.061458  0.277334      0  
4  0.502292  0.219422  0.215153  0.006685      0  

[5 rows x 31 columns]

In [8]:
data=data.drop_duplicates()

In [9]:
data.describe()

Time          V1          V2          V3          V4          V5  \
count  765.000000  765.000000  765.000000  765.000000  765.000000  765.000000   
mean   284.857516   -0.174935    0.211289    0.877483    0.280936   -0.007175   
std    171.463677    1.293200    1.174929    1.035273    1.262924    1.102495   
min      0.000000   -6.093248  -12.114213   -5.694973   -4.657545   -6.631951   
25%    130.000000   -0.895224   -0.182609    0.309867   -0.483154   -0.538265   
50%    284.000000   -0.378602    0.285832    0.907209    0.384451   -0.119566   
75%    433.000000    1.111048    0.885570    1.541254    1.126665    0.460941   
max    581.000000    1.586093    5.267376    3.772857    4.075817    7.672544   

               V6          V7          V8          V9  ...         V21  \
count  765.000000  765.000000  765.000000  765.000000  ...  765.000000   
mean     0.150284    0.126427   -0.059422   -0.032147  ...    0.001232   
std      1.224859    0.851834    0.833429    0.881308  ...    0.608942   
min     -3.498447   -4.925568   -7.494658   -2.770089  ...   -4.134608   
25%     -0.643710   -0.296178   -0.167739   -0.520297  ...   -0.215560   
50%     -0.117241    0.118066    0.032525   -0.095639  ...   -0.077543   
75%      0.474908    0.575308    0.251256    0.412254  ...    0.092532   
max      5.122103    4.808426    2.134599    5.459274  ...    5.273420   

              V22         V23         V24         V25         V26         V27  \
count  765.000000  765.000000  765.000000  765.000000  765.000000  765.000000   
mean    -0.097560   -0.040335    0.004438    0.112620    0.025911    0.023937   
std      0.605404    0.360256    0.616825    0.430973    0.485086    0.302052   
min     -2.776923   -3.553381   -1.867208   -1.389079   -1.243924   -2.377933   
25%     -0.524586   -0.176836   -0.370415   -0.169525   -0.313338   -0.045545   
50%     -0.081334   -0.049081    0.097487    0.141429   -0.015554    0.023386   
75%      0.304199    0.072651    0.427113    0.420079    0.263744    0.113704   
max      1.574750    3.150413    1.215279    1.136720    3.087444    2.490503   

              V28      Amount       Class  
count  765.000000  765.000000  765.000000  
mean    -0.017582    0.003018    0.011765  
std      0.279506    1.004721    0.107896  
min     -2.735623   -0.347318    0.000000  
25%     -0.033302   -0.317021    0.000000  
50%      0.020990   -0.259917    0.000000  
75%      0.086989   -0.063620    0.000000  
max      1.575380   19.014566    1.000000  

[8 rows x 31 columns]

In [10]:
data['Class'].value_counts()

Class
0    756
1      9
Name: count, dtype: int64

# TRAINING MODEL ON IMBALANCED DATA SET

In [14]:
x=data.drop('Class', axis=1)
y=data['Class']

In [15]:
from sklearn.model_selection import train_test_split


In [16]:
x_train,x_test, y_train, y_test=train_test_split(x,y,test_size=0.3)

In [17]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import accuracy_score, f1_score 

In [18]:
model=DecisionTreeClassifier()

In [19]:
model.fit(x_train,y_train)

DecisionTreeClassifier()

In [20]:
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print(f"\n Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"\n f1_score:{f1}")


 Accuracy:0.991304347826087

 f1_score:0.0


**Here we are getting f1_score as zero which indicates that our data is imbalanced,
so to balance our data we will use** **5 sampling methods-**
1. SMOTE
2. Randon Over Sampling
3. Randon Under Sampling
4. Tomek Links Under-Sampling
5. NearMiss
   
   **Over 5 models**
1. Decision tree
2. Random forest
3. Logistic regression
4. Linear regression
5. SVM

In [21]:
data['Class'].value_counts()

Class
0    756
1      9
Name: count, dtype: int64

In [22]:
import pandas as pd

sampling_methods = ['SMOTE', 'ROS', 'RUS', 'TMK', 'NM']
models = ['Decision tree', 'Random forest', 'Logistic Regression', 'Linear regression', 'SVM']

# Create an empty DataFrame with the specified rows and columns
df = pd.DataFrame(index=models, columns=sampling_methods)

# 1. SMOTE

In [23]:
from imblearn.over_sampling import SMOTE

In [24]:
x=data.drop('Class', axis=1)
y=data['Class']

In [25]:
x_res,y_res =SMOTE().fit_resample(x,y)

In [26]:
y_res.value_counts()

Class
0    756
1    756
Name: count, dtype: int64

**here sampling is done with class 1 and class 0 samples having 756 records each**

In [27]:
x_train,x_test, y_train, y_test=train_test_split(x_res,y_res,test_size=0.3)

In [28]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score ,mean_squared_error

In [29]:
#decision tree
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print("Decision Tree:")
print(f"Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"f1_score:{f1}")
df.loc['Decision tree', 'SMOTE'] = accuracy


Decision Tree:
Accuracy:0.9625550660792952
f1_score:0.9646569646569647


In [30]:
#random foreest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(x_train, y_train)
rf_pred = rf_model.predict(x_test)
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred, pos_label=1)

print("Random Forest:")
print(f"Accuracy: {rf_accuracy}")
print(f"F1 Score: {rf_f1}\n")
df.loc['Random forest', 'SMOTE'] = rf_accuracy

Random Forest:
Accuracy: 0.9933920704845814
F1 Score: 0.9936575052854123



In [31]:
# Logistic Regression
log_reg_model = LogisticRegression(max_iter=100, random_state=42)
log_reg_model.fit(x_train, y_train)
log_reg_pred = log_reg_model.predict(x_test)
log_reg_accuracy = accuracy_score(y_test, log_reg_pred)
log_reg_f1 = f1_score(y_test, log_reg_pred, pos_label=1)

print("Logistic Regression:")
print(f"Accuracy: {log_reg_accuracy}")
print(f"F1 Score: {log_reg_f1}\n")
df.loc['Logistic Regression', 'SMOTE'] = log_reg_accuracy

Logistic Regression:
Accuracy: 0.9118942731277533
F1 Score: 0.9180327868852459



C:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [32]:
# Linear Regression (for regression tasks)
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)
lin_reg_pred = lin_reg_model.predict(x_test)
lin_reg_mse = mean_squared_error(y_test, lin_reg_pred)

print("Linear Regression:")
print(f"Mean Squared Error: {lin_reg_mse}\n")

Linear Regression:
Mean Squared Error: 0.12228250748712269



In [33]:
# Support Vector Machine (SVM)
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train, y_train)
svm_pred = svm_model.predict(x_test)
svm_accuracy = accuracy_score(y_test, svm_pred)
svm_f1 = f1_score(y_test, svm_pred, pos_label=1)

print("Support Vector Machine:")
print(f"Accuracy: {svm_accuracy}")
print(f"F1 Score: {svm_f1}")
df.loc['SVM', 'SMOTE'] = svm_accuracy

Support Vector Machine:
Accuracy: 0.933920704845815
F1 Score: 0.9404761904761905


# 2. Random Over Sampling 

In [34]:
# import library
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler()
x_ros, y_ros = ros.fit_resample(x, y)

In [35]:
y_ros.value_counts()

Class
0    756
1    756
Name: count, dtype: int64

In [36]:
x_train,x_test, y_train, y_test=train_test_split(x_ros,y_ros,test_size=0.2)

In [37]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score ,mean_squared_error

In [38]:
#decision tree
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print("Decision Tree:")
print(f"Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"f1_score:{f1}")
df.loc['Decision tree', 'ROS'] = accuracy

Decision Tree:
Accuracy:0.9867986798679867
f1_score:0.987012987012987


In [39]:
#random foreest
rf_model = RandomForestClassifier()
rf_model.fit(x_train, y_train)
rf_pred = rf_model.predict(x_test)
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred, pos_label=1)

print("Random Forest:")
print(f"Accuracy: {rf_accuracy}")
print(f"F1 Score: {rf_f1}\n")
df.loc['Random forest', 'ROS'] = rf_accuracy

Random Forest:
Accuracy: 1.0
F1 Score: 1.0



In [40]:
# Logistic Regression
log_reg_model = LogisticRegression(max_iter=100, random_state=42)
log_reg_model.fit(x_train, y_train)
log_reg_pred = log_reg_model.predict(x_test)
log_reg_accuracy = accuracy_score(y_test, log_reg_pred)
log_reg_f1 = f1_score(y_test, log_reg_pred, pos_label=1)

print("Logistic Regression:")
print(f"Accuracy: {log_reg_accuracy}")
print(f"F1 Score: {log_reg_f1}\n")
df.loc['Logistic Regression', 'ROS'] = log_reg_accuracy

Logistic Regression:
Accuracy: 0.9141914191419142
F1 Score: 0.9212121212121213



C:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [41]:
# Linear Regression (for regression tasks)
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)
lin_reg_pred = lin_reg_model.predict(x_test)
lin_reg_mse = mean_squared_error(y_test, lin_reg_pred)

print("Linear Regression:")
print(f"Mean Squared Error: {lin_reg_mse}\n")


Linear Regression:
Mean Squared Error: 0.13160369821421597



In [42]:
# Support Vector Machine (SVM)
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train, y_train)
svm_pred = svm_model.predict(x_test)
svm_accuracy = accuracy_score(y_test, svm_pred)
svm_f1 = f1_score(y_test, svm_pred, pos_label=1)

print("Support Vector Machine:")
print(f"Accuracy: {svm_accuracy}")
print(f"F1 Score: {svm_f1}")
df.loc['SVM', 'ROS'] = svm_accuracy

Support Vector Machine:
Accuracy: 0.9108910891089109
F1 Score: 0.918429003021148


# 3. Random Under Sampling

In [43]:
x=data.drop('Class', axis=1)
y=data['Class']

In [44]:
# import library
from imblearn.under_sampling import RandomUnderSampler

rus = RandomUnderSampler(random_state=42, replacement=True)# fit predictor and target variable
x_rus, y_rus = rus.fit_resample(x, y)


In [45]:
x_train,x_test, y_train, y_test=train_test_split(x_rus,y_rus,test_size=0.2)

In [46]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score ,mean_squared_error

In [47]:
#decision tree
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print("Decision Tree:")
print(f"Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"f1_score:{f1}")
df.loc['Decision tree', 'RUS'] = accuracy


Decision Tree:
Accuracy:0.5
f1_score:0.5


In [48]:
#random forest
rf_model = RandomForestClassifier()
rf_model.fit(x_train, y_train)
rf_pred = rf_model.predict(x_test)
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred, pos_label=1)

print("Random Forest:")
print(f"Accuracy: {rf_accuracy}")
print(f"F1 Score: {rf_f1}\n")
df.loc['Random forest', 'RUS'] = rf_accuracy

Random Forest:
Accuracy: 0.5
F1 Score: 0.6666666666666666



In [49]:
# Logistic Regression
log_reg_model = LogisticRegression(max_iter=100, random_state=42)
log_reg_model.fit(x_train, y_train)
log_reg_pred = log_reg_model.predict(x_test)
log_reg_accuracy = accuracy_score(y_test, log_reg_pred)
log_reg_f1 = f1_score(y_test, log_reg_pred, pos_label=1)

print("Logistic Regression:")
print(f"Accuracy: {log_reg_accuracy}")
print(f"F1 Score: {log_reg_f1}\n")
df.loc['Logistic Regression', 'RUS'] = log_reg_accuracy

Logistic Regression:
Accuracy: 0.75
F1 Score: 0.8



C:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [50]:
# Linear Regression (for regression tasks)
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)
lin_reg_pred = lin_reg_model.predict(x_test)
lin_reg_mse = mean_squared_error(y_test, lin_reg_pred)

print("Linear Regression:")
print(f"Mean Squared Error: {lin_reg_mse}\n")


Linear Regression:
Mean Squared Error: 0.41364710964308676



In [51]:
# Support Vector Machine (SVM)
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train, y_train)
svm_pred = svm_model.predict(x_test)
svm_accuracy = accuracy_score(y_test, svm_pred)
svm_f1 = f1_score(y_test, svm_pred, pos_label=1)

print("Support Vector Machine:")
print(f"Accuracy: {svm_accuracy}")
print(f"F1 Score: {svm_f1}")
df.loc['SVM', 'RUS'] = svm_accuracy

Support Vector Machine:
Accuracy: 0.75
F1 Score: 0.8


# 4. Tomek Links Under Sampling

In [52]:
x=data.drop('Class', axis=1)
y=data['Class']

In [53]:
# import library
from imblearn.under_sampling import TomekLinks

tl = RandomOverSampler(sampling_strategy='majority')

# fit predictor and target variable
x_tl, y_tl = ros.fit_resample(x, y)


In [54]:
x_train,x_test, y_train, y_test=train_test_split(x_tl,y_tl,test_size=0.2)

In [55]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score ,mean_squared_error

In [56]:
#decision tree
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print("Decision Tree:")
print(f"Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"f1_score:{f1}")
df.loc['Decision tree', 'TMK'] = accuracy

Decision Tree:
Accuracy:0.9867986798679867
f1_score:0.9866666666666667


In [57]:
#random foreest
rf_model = RandomForestClassifier()
rf_model.fit(x_train, y_train)
rf_pred = rf_model.predict(x_test)
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred, pos_label=1)

print("Random Forest:")
print(f"Accuracy: {rf_accuracy}")
print(f"F1 Score: {rf_f1}\n")
df.loc['Random forest', 'TMK'] = rf_accuracy

Random Forest:
Accuracy: 1.0
F1 Score: 1.0



In [58]:
# Logistic Regression
log_reg_model = LogisticRegression(max_iter=100, random_state=42)
log_reg_model.fit(x_train, y_train)
log_reg_pred = log_reg_model.predict(x_test)
log_reg_accuracy = accuracy_score(y_test, log_reg_pred)
log_reg_f1 = f1_score(y_test, log_reg_pred, pos_label=1)

print("Logistic Regression:")
print(f"Accuracy: {log_reg_accuracy}")
print(f"F1 Score: {log_reg_f1}\n")
df.loc['Logistic Regression', 'TMK'] = log_reg_accuracy

Logistic Regression:
Accuracy: 0.9075907590759076
F1 Score: 0.9135802469135802



C:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [59]:
# Linear Regression (for regression tasks)
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)
lin_reg_pred = lin_reg_model.predict(x_test)
lin_reg_mse = mean_squared_error(y_test, lin_reg_pred)

print("Linear Regression:")
print(f"Mean Squared Error: {lin_reg_mse}\n")

Linear Regression:
Mean Squared Error: 0.15052320768606522



In [60]:
# Support Vector Machine (SVM)
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train, y_train)
svm_pred = svm_model.predict(x_test)
svm_accuracy = accuracy_score(y_test, svm_pred)
svm_f1 = f1_score(y_test, svm_pred, pos_label=1)

print("Support Vector Machine:")
print(f"Accuracy: {svm_accuracy}")
print(f"F1 Score: {svm_f1}")
df.loc['SVM', 'TMK'] = svm_accuracy

Support Vector Machine:
Accuracy: 0.9042904290429042
F1 Score: 0.9107692307692308


# 5.Near Miss

In [61]:
x=data.drop('Class', axis=1)
y=data['Class']

In [62]:
from imblearn.under_sampling import NearMiss
nm = NearMiss()
x_nm, y_nm = nm.fit_resample(x, y)


In [63]:
x_train,x_test, y_train, y_test=train_test_split(x_nm,y_nm,test_size=0.3)

In [64]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score ,mean_squared_error

In [65]:
#decision tree
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
accuracy= accuracy_score(y_test,y_pred)
print("Decision Tree:")
print(f"Accuracy:{accuracy}")
f1= f1_score(y_test,y_pred)
print(f"f1_score:{f1}")
df.loc['Decision tree', 'NM'] = accuracy

Decision Tree:
Accuracy:0.5
f1_score:0.4


In [66]:
#random foreest
rf_model = RandomForestClassifier()
rf_model.fit(x_train, y_train)
rf_pred = rf_model.predict(x_test)
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred, pos_label=1)

print("Random Forest:")
print(f"Accuracy: {rf_accuracy}")
print(f"F1 Score: {rf_f1}\n")
df.loc['Random forest', 'NM'] = rf_accuracy

Random Forest:
Accuracy: 0.6666666666666666
F1 Score: 0.6666666666666666



In [67]:
# Logistic Regression
log_reg_model = LogisticRegression(max_iter=100, random_state=42)
log_reg_model.fit(x_train, y_train)
log_reg_pred = log_reg_model.predict(x_test)
log_reg_accuracy = accuracy_score(y_test, log_reg_pred)
log_reg_f1 = f1_score(y_test, log_reg_pred, pos_label=1)

print("Logistic Regression:")
print(f"Accuracy: {log_reg_accuracy}")
print(f"F1 Score: {log_reg_f1}\n")
df.loc['Logistic Regression', 'NM'] = log_reg_accuracy

Logistic Regression:
Accuracy: 0.5
F1 Score: 0.4



C:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [68]:
# Linear Regression (for regression tasks)
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)
lin_reg_pred = lin_reg_model.predict(x_test)
lin_reg_mse = mean_squared_error(y_test, lin_reg_pred)

print("Linear Regression:")
print(f"Mean Squared Error: {lin_reg_mse}\n")

Linear Regression:
Mean Squared Error: 0.4064991192073733



In [69]:
# Support Vector Machine (SVM)
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train, y_train)
svm_pred = svm_model.predict(x_test)
svm_accuracy = accuracy_score(y_test, svm_pred)
svm_f1 = f1_score(y_test, svm_pred, pos_label=1)

print("Support Vector Machine:")
print(f"Accuracy: {svm_accuracy}")
print(f"F1 Score: {svm_f1}")
df.loc['SVM', 'NM'] = svm_accuracy

Support Vector Machine:
Accuracy: 0.5
F1 Score: 0.4


In [ ]:
max_value = df.max().max()
max_method = df.max(axis=0).idxmax()  # Sampling method (column)
max_model = df[max_method].idxmax()  # Model name (row)

# Print the results
print(f"The maximum value is {max_value} found in model '{max_model}' with sampling method '{max_method}'.")